### ISSUE TITLE
Creating valid HQDM Things

### ISSUE THREAD

**Issue by twalmsley** [gchq/MagmaCore#50]
Prior to the recent rewrite the HQDM objects had some support for creating valid HQDM objects, but this was RDF-specific and was removed so that HQDM can be used without requiring the use of RDF.

There is now a gap in functionality for validating models produced using MagmaCore, and although the `DataIntegrityReport` feature goes some way towards helping this, it is checking _after_ the data has been persisted and is more useful as part of system testing.

Ideally there would be some type-safe mechanism for checking that programs will produce minimally valid objects at compile time, and then the `DataIntegrityReport` can help with other issues later in the development cycle.

Even better, but more difficult, would be a type-safe way of building entire object structures for checking at compile time - this would make it easier for errors to be prevented early in the development cycle rather than being detected and corrected later.

**Comment by twalmsley** [comment]
Working on a solution proposal - to be discussed before acceptance/rejection.

**Comment by twalmsley** [comment]
This proposal is an attempt to make it easier for developers to know what are the possible relationships that entities can have, but is starting to look messy. First of all it's a lot of work, and here is an extreme example:
```java
    public static AcceptanceOfOfferForGoods createAcceptanceOfOfferForGoods(
            final String id,
            final Event beginning,
            final Event ending,
            final Set<SpatioTemporalExtent> aggregatedInto,
            final Set<Event> causes,
            final Set<SpatioTemporalExtent> consistsOf,
            final Set<Activity> consistsOfActivities,
            final Set<Participant> consistsOfParticipant,
            final Set<Thing> determines,
            final Set<Class> memberOfClass,
            final Set<ClassOfSociallyConstructedActivity> memberOf,
            final Set<KindOfActivity> memberOfKind,
            final Set<SpatioTemporalExtent> partOf,
            final Set<OfferAndAcceptanceForGoods> partOfOfferAndAcceptanceForGoods,
            final Set<AgreementExecution> partOfAgreementExecutions,
            final Set<PossibleWorld> partOfPossibleWorlds,
            final Set<OfferForGoods> references,
            final Set<SpatioTemporalExtent> temporalPartOf,
            final Set<Individual> temporalPartOfIndividuals) {
// ...
    }
```
I believe that in most client code, the majority of these parameters would be `null` so this doesn't look like a good option. 

Another possible option I'll look into is to re-introduce the RDF-specific `Builder` classes outside of the `hqdm` module.

**Comment by twalmsley** [comment]
Waiting on som PRs so I can rebase this change and create another PR.
